### PR TITLE
4_4_deep_comparison

### PR DESCRIPTION
**Pull request by TheIronHobo**
### Deep Comparison
- Performs deep comparison of two types and determines equality of enumerable data types
- Order of properties and values no longer matter
- Nested objects/arrays are also compared
- Returns true if both objects have no enumerable data types

### CODE REVIEW

File: eloquentjs/4_4_deep_comparison.js [hunk 1]
```diff
@@ -0,0 +1,71 @@
+/**
+ * Performs deep comparison of two types and determines equality
+ * @param {*} itemA 
+ * @param {*} itemB 
+ * @returns 
+ */
+function deepEqual(itemA, itemB) {
+    if(itemA === null && itemB === null) {
```
Comment (CalHinshaw) [inline eloquentjs/4_4_deep_comparison.js]: Perhaps this should just be if itemA === itemB return true? Referential equality is certainly deep equality - it's the same molecules in the computer!

File: eloquentjs/4_4_deep_comparison.js [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+        return true;
+    }
+    if (typeof itemA === typeof itemB) {
```
Comment (CalHinshaw) [inline eloquentjs/4_4_deep_comparison.js]: It's best to try to make your conditions as un-nested as possible. if typeof a !== typeof b you can return false from the whole function which lets you unnest the inner contents

File: eloquentjs/4_4_deep_comparison.js [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+
+            if (keysA.length !== keysB.length) {
+                return false;
+            }
+
+            if (keysA.length === 0) {
+                throw "objects being compared have no enumerable data types";
+            }
+            
+            for (let i = 0; i < keysA.length; i++) {
```
Comment (CalHinshaw) [inline eloquentjs/4_4_deep_comparison.js]: The ordering behavior described in [this](https://dev.to/frehner/the-order-of-js-object-keys-458d) article applies to Object.entries as well. Can you figure out what's wrong with your implementation?

File: eloquentjs/4_4_deep_comparison.js [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+                let keyValuePairA = keysA[i];
+                let keyValuePairB = keysB[i];
+                if (deepEqual(keyValuePairA[0], keyValuePairB[0]) === false || deepEqual(keyValuePairA[1], keyValuePairB[1]) === false) {
+                    return false;
+                }
+            }
+
+            return true;
+        } else {
+            return (itemA === itemB) ? true : false;
```
Comment (CalHinshaw) [inline eloquentjs/4_4_deep_comparison.js]: no need for a ternary, you can just use the result of a boolean expression directly

File: eloquentjs/4_4_deep_comparison.js [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+        }
+    }
+    else { 
+        return false;
+    }
+}
+
+let jimFavorites    = {favoriteNumber: 3, favoriteFruit: 'apple', favoriteDog: {name: 'simon', breed: 'border collie'}, favoriteColors: ['red', 'blue']};
```
Comment (CalHinshaw) [inline eloquentjs/4_4_deep_comparison.js]: You're missing some important test cases. The linked article may give you enough info to come up with them. See if you can!

**Review by CalHinshaw — CHANGES_REQUESTED** [review]
This is a great start to a subtly hard problem, great job!

File: eloquentjs/4_4_deep_comparison.js (hunk 3 above)
Comment (TheIronHobo) [reply eloquentjs/4_4_deep_comparison.js]: Oh beans. That's pretty devious. If I found the same problem I believe its this : 
```
let testCarA = {3: ['a','b','c']}; // numerical 3
let testCarB = {'3': ['a','b','c']}; // string 3
console.log(deepEqual(testCarA, testCarB)); // returns true 🤦‍♂️
```
Implementing a guard for this in the next push.

File: eloquentjs/4_4_deep_comparison.js [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+ * @returns 
+ */
+function deepEqual(itemA, itemB) {
+    if(itemA === itemB) {
+        return true;
+    }
+    if (typeof itemA !== typeof itemB) {
+        return false;
+    }
+    if(itemA === null || itemB === null) {
```
Comment (CalHinshaw) [inline eloquentjs/4_4_deep_comparison.js]: space after if

File: eloquentjs/4_4_deep_comparison.js [hunk 7]
```diff
@@ -0,0 +1,84 @@
+/**
+ * Performs deep comparison of two inputs and determines equality
+ * @param {*} itemA 
+ * @param {*} itemB 
+ * @returns 
+ */
+function deepEqual(itemA, itemB) {
+    if(itemA === itemB) {
```
Comment (CalHinshaw) [inline eloquentjs/4_4_deep_comparison.js]: space after if

File: eloquentjs/4_4_deep_comparison.js [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+
+    if (entriesA.length !== entriesB.length) {
+        return false;
+    }
+
+    entriesA.sort();
+    entriesB.sort();
+
+    for (let i = 0; i < entriesA.length; i++) {
+        let keyValuePairA = entriesA[i];
```
Comment (CalHinshaw) [inline eloquentjs/4_4_deep_comparison.js]: Don't use let for constant values

File: eloquentjs/4_4_deep_comparison.js [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+    }
+
+    let entriesA = Object.entries(itemA);
+    let entriesB = Object.entries(itemB);
+
+    if (entriesA.length !== entriesB.length) {
+        return false;
+    }
+
+    entriesA.sort();
```
Comment (CalHinshaw) [inline eloquentjs/4_4_deep_comparison.js:28]: This is a good enough solution for now, but you can actually do it in linear time by checking to make sure they both have the same number of keys and then iterating overt the keys in one and using constant time access into the other object to make sure they have equal values.

Also, just for your understanding, sort may work in a way you're not expecting: https://replit.com/@SKLARGBLAR/objkeysexampe#index.js

File: eloquentjs/4_4_deep_comparison.js [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+    entriesB.sort();
+
+    for (let i = 0; i < entriesA.length; i++) {
+        let keyValuePairA = entriesA[i];
+        let keyValuePairB = entriesB[i];
+
+        let keysEqual = deepEqual(keyValuePairA[0], keyValuePairB[0]);
+        let valuesEqual = deepEqual(keyValuePairA[1], keyValuePairB[1])
+
+        if (!keysEqual) {
```
Comment (CalHinshaw) [inline eloquentjs/4_4_deep_comparison.js]: these two if statements could be combined to be more clear

File: eloquentjs/4_4_deep_comparison.js [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+        if (!valuesEqual) {
+            return false;
+        }
+    }
+
+    return true;
+}
+
+let jimFavorites    = {favoriteNumber: 3, favoriteFruit: 'apple', favoriteDog: {name: 'simon', breed: 'border collie'}, favoriteColors: ['red', 'blue']};
+let janeFavorites   = {favoriteNumber: 3, favoriteFruit: 'apple', favoriteDog: {name: 'simon', breed: 'border collie'}, favoriteColors: ['red', 'blue']}; // Same as jim
```
Comment (CalHinshaw) [inline eloquentjs/4_4_deep_comparison.js]: The way these tests are organized is a bit confusing so I'm not sure, but it seems like you're not testing deep equals on objects with different insertion orders. I think you should add some cases like that and try to make the way you've split these cases up clearer.

**Review by CalHinshaw — CHANGES_REQUESTED** [review]
Looking way better, great work!

**Review by CalHinshaw — APPROVED** [review]
Looks great, good job!

**Pull request merged**